File: app/src/main/java/se/creotec/chscardbalance2/service/MenuService.java
// Copyright (c) 2017 Alexander Håkansson
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT
package se.creotec.chscardbalance2.service;

import android.content.Intent;
import android.support.annotation.Nullable;
import android.util.Log;

import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;

import se.creotec.chscardbalance2.Constants;
import se.creotec.chscardbalance2.model.BackendResponse;
import se.creotec.chscardbalance2.model.MenuData;

public class MenuService extends BackendService<MenuData> {

    private static final String LOG_TAG = MenuService.class.getName();
    private final Type responseType = new TypeToken<BackendResponse<MenuData>>() {}.getType();


    public MenuService() {
        super(MenuService.class.getName());
    }

    @Override
    protected void onHandleIntent(@Nullable Intent intent) {
        Log.i(LOG_TAG, "Received intent");
        if (intent == null || intent.getAction() == null) {
            return;
        }
        if (intent.getAction().equals(Constants.ACTION_UPDATE_MENU)) {
            try {
                BackendResponse<MenuData> response = getBackendData(Constants.ENDPOINT_MENU, Constants.ENDPOINT_MENU_LANG_EN);
                Log.i(LOG_TAG, "Got response: " + response.getData().toString());
            } catch (BackendFetchException e) {
                Log.e(LOG_TAG, e.getMessage());
            }
        }
    }

    @Override
    protected void validateVariable(String variable) throws BackendFetchException {
        switch (variable) {
            case Constants.ENDPOINT_MENU_LANG_EN:
            case Constants.ENDPOINT_MENU_LANG_SV:
                break;
            default:
                throw new BackendFetchException("Unsupported language: " + variable);
        }
    }

    @Override
    protected Type getResponseType() {
        return responseType;
    }
}

File: app/src/main/java/se/creotec/chscardbalance2/service/MenuService.kt
// Copyright (c) 2017 Alexander Håkansson
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT
package se.creotec.chscardbalance2.service

import android.content.Intent
import android.util.Log

import com.google.gson.reflect.TypeToken

import java.lang.reflect.Type

import se.creotec.chscardbalance2.Constants
import se.creotec.chscardbalance2.GlobalState
import se.creotec.chscardbalance2.model.BackendResponse
import se.creotec.chscardbalance2.model.MenuData


class MenuService : AbstractBackendService<MenuData>(MenuService::class.java.name) {
    private val responseType = object : TypeToken<BackendResponse<MenuData>>() {}.type
    private val LOG_TAG = MenuService::class.java.name

    override fun onHandleIntent(intent: Intent?) {
        Log.i(LOG_TAG, "Received intent")
        if (intent == null || intent.action == null) {
            return
        }
        if (intent.action == Constants.ACTION_UPDATE_MENU) {
            try {
                val global = application as GlobalState
                val language = global.model.preferredMenuLanguage
                val response = getBackendData(Constants.ENDPOINT_MENU, language)
                global.model.menuData = response.data
                global.saveMenuData()
                Log.i(LOG_TAG, "Got response: " + response.data.toString())
            } catch (e: BackendFetchException) {
                Log.e(LOG_TAG, e.message)
            }

        }
    }

    @Throws(BackendFetchException::class)
    override fun validateVariable(variable: String) {
        when (variable) {
            Constants.ENDPOINT_MENU_LANG_EN, Constants.ENDPOINT_MENU_LANG_SV -> {
            }
            else -> throw BackendFetchException("Unsupported language: " + variable)
        }
    }

    override fun getResponseType(): Type {
        return responseType
    }
}
